packages/flutter_tools/lib/src/build_system/build_system.dart

```diff
@@ -0,0 +1,574 @@
+// found in the LICENSE file.
+
+import 'dart:convert';
+
+import 'package:crypto/crypto.dart';
+import 'package:meta/meta.dart';
+
+import '../base/file_system.dart';
+import '../base/platform.dart';
+import '../build_info.dart';
+import '../cache.dart';
+import '../convert.dart';
+import '../globals.dart';
+import 'targets.dart';
+
+/// Return the name for the build mode, or "any" if null.
+String getNameForBuildMode(BuildMode buildMode) {
+  switch (buildMode) {
+    case BuildMode.debug:
+      return 'debug';
+    case BuildMode.profile:
+      return 'profile';
+    case BuildMode.release:
+      return 'relese';
+    case BuildMode.dynamicProfile:
+      return 'dynamic-profile';
+    case BuildMode.dynamicRelease:
+      return 'dynamic-release';
+  }
+  return 'any';
+}
+
+/// Returns the [BuildMode] for a particular `name`.
+BuildMode getBuildModeForName(String name) {
+  switch (name) {
+    case 'debug':
+      return BuildMode.debug;
+    case 'profile':
+      return BuildMode.profile;
+    case 'release':
+      return BuildMode.release;
+    default:
+      throw ArgumentError('$name is not a valid build mode');
+  }
+}
+
+/// Returns the host folder name for a particular target platform.
+String getHostFolderForTargetPlaltform(TargetPlatform targetPlatform) {
+  switch (targetPlatform) {
+    case TargetPlatform.android_arm:
+    case TargetPlatform.android_arm64:
+    case TargetPlatform.android_x64:
+    case TargetPlatform.android_x86:
+      return 'android';
+    case TargetPlatform.ios:
+      return 'ios';
+    case TargetPlatform.darwin_x64:
+      return 'macos';
+    case TargetPlatform.linux_x64:
+      return 'linux';
+    case TargetPlatform.windows_x64:
+      return 'windows';
+    case TargetPlatform.fuchsia:
+      return 'fuchsia';
+    case TargetPlatform.web:
+      return 'web';
+    case TargetPlatform.tester:
+      throw UnsupportedError('tester is not a support platform for building.');
+  }
+  return 'any';
+}
+
+
+/// An input function produces a list of additional input files for an
+/// environment.
+typedef InputFunction = List<FileSystemEntity> Function(Environment environment);
+
+/// An exception thrown when a rule declares an output that was not produced
+/// by the invocation.
+class MissingOutputException implements Exception {
+  const MissingOutputException(this.file, this.target);
+
+  /// The file we expected to find.
+  final File file;
+
+  /// The name of the target this file should have been output from.
+  final String target;
+
+  @override
+  String toString() {
+    return '${file.path} was declared as an output, but was not generated by '
+    'the invocation. Check the definition of target:$target for errors';
+  }
+}
+
+/// Finds the locations of all dart files within the project.
+///
+/// This does not attempt to determine if a file is used or imported, so it
+/// may otherwise report more files than strictly necessary.
+List<FileSystemEntity> listDartSources(
```

Comment:
This is somewhat of a WIP

Reply:
I imagine that we would skip the deps file approach, and instead use logic similar to the hot reload: all dart files in this project, the pubspec file, and all dart files from path dependencies.

This is a larger set than what the depfile gives us, but we can know it ahead of time